runtime/lua/vim/keymap.lua

```diff
@@ -0,0 +1,127 @@
+---     return vim.fn.pumvisible() == 1 and "<C-n>" or "<Tab>"
+---   end, { silent = true, expr = true, replace_termcodes = true })
+---   -- While setting <Plug> mapping remember to set remap to true
+---   vim.keymap.set('n', '[%', '<Plug>(MatchitNormalMultiBackward)', {silent = true, remap = true })
+--- </pre>
+--@param mode (string|table):   Same mode short names as |nvim_set_keymap|.
+---                             Can also be list of modes to create keymap on multiple modes.
+--@param lhs (string):          Left-hand-side |{lhs}| of the mapping.
+--@param rhs (string|function): Right-hand-side |{rhs}| of the mapping. Can also be a lua function.
+--
+--@param opts (table): A table of |:map-arguments| such as "silent". See |nvim_set_keymap()|
+---       along with "replace_termcodes". When replace_termcodes is true with
+---       lua function expr maps |nvim_replace_termcodes| is applied to return
+---       value of mapped function.
+---       One difference compared to |nvim_set_keymap| is here "remap" used as option
+---       instead of |noremap| . It's meaning is opposite of |noremap| and it's false by default.
+--@see |nvim_set_keymap|
+---
+function keymap.set(modes, lhs, rhs, opts)
+  opts = opts or {}
+
+  vim.validate {
+    mode = {modes, {'s', 't'}},
+    lhs = {lhs, 's'},
+    rhs = {rhs, {'s', 'f'}},
+    opts = {opts, 't'}
+  }
+
+  local buffer = false
+  local is_rhs_lua = type(rhs) == "function"
+  modes = type(modes) == 'string' and {modes} or modes
+
+  if opts.buffer ~= nil then
+    buffer = opts.buffer == true and 0 or opts.buffer
+    opts.buffer = nil
+  end
+
+  if opts.replace_termcodes ~= nil then
+    if opts.replace_termcodes and is_rhs_lua then
+      local user_rhs = rhs
+      rhs = function ()
+        return vim.api.nvim_replace_termcodes(user_rhs(), true, true, true)
+      end
+    end
+    opts.replace_termcodes = nil
+  end
+
+  -- remap defaults to false
+  opts.noremap = true
+  if opts.remap ~= nil then
+    opts.noremap = not opts.remap
+    opts.remap = nil
+  end
+
+  if is_rhs_lua then
+    opts.callback = rhs
+    rhs = ''
+  end
+
+  if buffer == false then
+    for _, mode in ipairs(modes) do
+      vim.api.nvim_set_keymap(mode, lhs, rhs, opts)
+    end
+  else
+    for _, mode in ipairs(modes) do
+      vim.api.nvim_buf_set_keymap(buffer, mode, lhs, rhs, opts)
+    end
+  end
+end
+
+--- Remove an existing keymap.
+--- Examples:
+--- <pre>
+---   vim.keymap.del('n', 'lhs')
+---
+---   vim.keymap.del({'n', 'i', 'v'}, '<leader>w', { buffer = 5 })
+--- </pre>
+--@param opts (table): A table optional arguments:
+---     - buffer: Can be boolean or bufnr . Needed when removing buffer
+---               local keymaps. buffer true or 0 refers to current buffer.
+--@see |vim.keymap.set|
+---
+function keymap.del(modes, lhs, opts)
+  opts = opts or {}
+
+  vim.validate {
+    mode = {modes, {'s', 't'}},
+    lhs = {lhs, 's'},
+    opts = {opts, 't'}
+  }
```

Comment:
Same thing here, make `vim.validate` the first thing and make the `opts` argument optional, and put `opts = opts or {}` below `vim.validate`